### ISSUE TITLE
Small Question About Truncation

### ISSUE THREAD

**Issue by Orion-Zheng** [NormXU/Consistent-DynamicNTKRoPE#1]
<img width="824" alt="image" src="https://github.com/NormXU/Consistent-DynamicNTKRoPE/assets/62330719/5f5c22ed-7cae-4b55-bd6d-675f36166fd1">
Did you ignore losses of some tokens at the beginning? If yes, may I ask how many tokens did you ignore?  

When trying to implement this on my own, I found that the losses at beginning time steps are very high so it will affect the results when taking the average of the whole sequnece losses. I think this is the reason we need to truncate some losses at the beginning. Am I correct?😃  

Besides I notice that to compute the loss of different time steps, your code truncate samples to multiple time to get avg ppl at different length. I think this can be optimized to only compute once by use the cumulative sum. For example we have losses of n positions given logits and labels after one forward:  

[loss_1, loss_2, loss_3, ..., loss_n] --torch.cumsum()--> [loss_1, loss_1+2,  loss_1+2+3, ..., loss_1+2+3+...+n] 
-=/(torch.arange(seq_len)+1)-> [mean(loss_1), mean(loss_1+2),  mean(loss_1+2+3), ..., mean(loss_1+2+3+...+n)]  
-exp()->  [avg_ppl_1, avg_ppl_2, avg_ppl_3, ..., avg_ppl_n]  

Do you think this is a correct implementation?😃

**Comment by NormXU** [comment]
@Orion-Zheng Your understanding regarding why I skipped some tokens is correct. 

If the small number of tokens the ppl are computed based on doesn't form a coherent sequence, the ppl value will be very large. Therefore, it cannot reflect the model's typical performance over a certain dataset. That is why I started from 200 tokens to compute the first ppl value.

As for the loss optimization, if I get your idea correctly, you are trying to compute CrossEntropyLoss between $N$ input labels and $N$ output logits. Let's say $m < N$, and you denote $loss_m = \text{CrossEntropy}(labels[:m]，logits[:m] )$.

However, this is not a correct implementation.  In language models, the $ppl_N$ is used to describe how well the probability distribution predicted by the model aligns with the actual distribution of the words given the previous context whose length is $N$.  

Therefore, $ppl_m \neq \text{CrossEntropy}(labels[:m]，logits[:m] )$ since token $m$ can see bi-directional in your implementation, but we expect the mth token should not see the tokens after it when computing $ppl_m$,
